DataFormats/FTLRecHit/interface/FTLCluster.h

```diff
@@ -0,0 +1,272 @@
+  
+  
+  inline float energy() const {
+    float qm = 0;
+    int isize = theHitENERGY.size();
+    for (int i=0; i<isize; ++i) 
+      qm += theHitENERGY[i];
+    return qm;
+  } // Return total cluster energy.
+  
+  inline int minHitRow() const { return theMinHitRow;} // The min x index.
+  inline int maxHitRow() const { return minHitRow() + rowSpan();} // The max x index.
+  inline int minHitCol() const { return theMinHitCol;} // The min y index.
+  inline int maxHitCol() const { return minHitCol() + colSpan();} // The max y index.
+  
+  const std::vector<uint8_t> & hitOffset() const { return theHitOffset;}
+  const std::vector<float> & hitENERGY() const { return theHitENERGY;}
+  const std::vector<float> & hitTIME() const { return theHitTIME;}
+  const std::vector<float> & hitTIME_ERROR() const { return theHitTIME_ERROR;}
+    
+  // infinite faster than above...
+  FTLHit hit(int i) const {
+    return FTLHit(minHitRow() + theHitOffset[i*2],
+		 minHitCol() + theHitOffset[i*2+1],
+		  theHitENERGY[i],
+		  theHitTIME[i],
+		  theHitTIME_ERROR[i]
+		 );
+
+  }
+  
+  FTLHit seed() const {
+    return hit(seed_);
+  }
+  
+  int colSpan() const {return theHitColSpan; }
+  
+  int rowSpan() const { return theHitRowSpan; }
+  
+  const DetId& id() const { return theid; }
+  const DetId& detid() const { return id(); }
+
+  bool overflowCol() const { return overflow_(theHitColSpan); }
+  
+  bool overflowRow() const { return overflow_(theHitRowSpan); }
+  
+  bool overflow() const { return  overflowCol() || overflowRow(); }
+  
+  void packCol(uint16_t ymin, uint16_t yspan) {
+    theMinHitCol = ymin;
+    theHitColSpan = std::min(yspan, uint16_t(MAXSPAN));
+  }
+  void packRow(uint16_t xmin, uint16_t xspan) {
+    theMinHitRow = xmin;
+    theHitRowSpan = std::min(xspan, uint16_t(MAXSPAN));
+  }
+  
+   void setClusterErrorX( float errx ) { err_x = errx; }
+   void setClusterErrorY( float erry ) { err_y = erry; }
+   void setClusterErrorTime( float errtime ) { err_time = errtime; }
+   float getClusterErrorX() const { return err_x; }
+   float getClusterErrorY() const { return err_y; }
+   float getClusterErrorTime() const { return err_time; }
+  
+private:
+
+   DetId theid;
+   
+   std::vector<uint8_t>  theHitOffset;
+   std::vector<float> theHitENERGY;
+   std::vector<float> theHitTIME;
+   std::vector<float> theHitTIME_ERROR;
+   
+   uint16_t theMinHitRow=MAXPOS; // Minimum hit index in the x direction (low edge).
+   uint16_t theMinHitCol=MAXPOS; // Minimum hit index in the y direction (left edge).
+   uint8_t theHitRowSpan=0; // Span hit index in the x direction (low edge).
+   uint8_t theHitColSpan=0; // Span hit index in the y direction (left edge).
+   
+   float err_x=-99999.9f;
+   float err_y=-99999.9f;
+   float err_time=-99999.9f;
+   
+   uint8_t seed_;
+   
+   float weighted_sum(const std::vector<float>& weights, const std::function<float (unsigned int i)>& sumFunc, const std::function<float (float,float)>& outFunc) const 
```

Comment:
Making this a template function would probably let the optimizer work much better.

Reply:
@Dr15Jones  I understand you suggest to replace the use of ```std::function``` with something like what discussed in https://stackoverflow.com/questions/16440029/c-function-passed-as-template-argument-vs-parameter , am I right?

Reply:
Yes, something like
```C++
template<typename SumFunc, typename OutFunc> float weighted_sum(const std::vector<float>& weights, SumFunc&& sumFunc, OutFunc&& outFunc) const;
```